metricbeat/module/system/process/process.go

```diff
@@ -111,6 +114,7 @@ func (m *MetricSet) Fetch(r mb.ReporterV2) error {
 	// monitor either a single PID, or the configured set of processes.
 	if m.setpid == 0 {
 		procs, roots, err := m.stats.Get()
+		m.UpdateStatusOnErr(err)
```

Comment:
Is there a reason we're implementing it this way, where the individual metricsets are responsible for setting an error state? It means we have to go through and add new code to all of the metricsets. `Fetch()` here returns an error, and that error is caught in `(msw *metricSetWrapper) fetch()` in `metricbeat/mb/module/wrapper.go`. If we add reporting there, any metricset that returns an error will have their state reported upstream to agent. Or we could modify the `Event(event mb.Event)` methods in `wrapper.go` to check if `mb.Event.Error` is set.

Reply:
I think it's because https://github.com/elastic/beats/pull/39209 did the same for `filebeat` for `input/cel`.

Reply:
@fearful-symmetry I think it's better to check for `mb.Event.Error` instead of updating every metricset.

It would be great if @cmacknz chimes in here.

Reply:
Yeah, agreed that we should do it in such a way that we don't need to implement something special for every individual metricset. Personally, I would say that it doesn't matter if we diverge too much from what filebeat does here, since filebeat and metricbeat are already pretty divergent.

Reply:
No opposition to catching errors in a generic way here. The Metricbeat metricset concept is much more narrowly focused in what it lets implementations do, almost everything happens on a schedule and we get the result of each attempt to collect data.

For Filebeat, catching errors at the top level would only catch failure to create the input, the v2/input interface gives no visibility into what happens during data collection the way Metricbeat does.

https://github.com/elastic/beats/blob/90719641eab8c7584648d0113bf374edb584217c/filebeat/input/v2/input.go#L58-L76